sources/rest_api/client.py

```diff
@@ -0,0 +1,252 @@
+from typing import Iterator, Optional, List, Dict, Any, TypeVar, Iterable, cast, Literal
+import copy
+from urllib.parse import urlparse
+
+from requests.auth import AuthBase
+from requests import Session as BaseSession
+from requests import Response, Request
+
+from dlt.common import logger
+from dlt.common import jsonpath
+from dlt.sources.helpers.requests.retry import Client
+
+from .typing import HTTPMethodBasic, HTTPMethod
+from .paginators import BasePaginator
+from .auth import AuthConfigBase
+from .detector import create_paginator, find_records
+
+from .utils import join_url
+
+
+_T = TypeVar("_T")
+
+
+class PageData(List[_T]):
+    """A list of elements in a single page of results with attached request context.
+
+      The context allows to inspect the response, paginator and authenticator, modify the request
+    """
+    def __init__(
+        self,
+        __iterable: Iterable[_T],
+        request: Request,
+        response: Response,
+        paginator: BasePaginator,
+        auth: AuthConfigBase
+    ):
+        super().__init__(__iterable)
+        self.request = request
+        self.response = response
+        self.paginator = paginator
+        self.auth = auth
+
+
+class RESTClient:
+    """A generic REST client for making requests to an API.
+
+    Attributes:
+        base_url (str): The base URL of the API.
+        headers (Optional[Dict[str, str]]): Headers to include in all requests.
+        auth (Optional[AuthBase]): An authentication object to use for all requests.
+        paginator (Optional[BasePaginator]): A paginator object for handling API pagination.
+            Note that this object will be deepcopied for each request to ensure that the
+            paginator state is not shared between requests.
+    """
+
+    def __init__(
+        self,
+        base_url: str,
+        headers: Optional[Dict[str, str]] = None,
+        auth: Optional[AuthConfigBase] = None,
+        paginator: Optional[BasePaginator] = None,
+        data_selector: Optional[jsonpath.TJsonPath] = None,
+        session: BaseSession = None,
+    ) -> None:
+        self.base_url = base_url
+        self.headers = headers
+        self.auth = auth
+
+        if session:
+            self._validate_session_raise_for_status(session)
+            self.session = session
+        else:
+            self.session = Client(raise_for_status=False).session
```

Comment:
I think we need to use `Client` obj as session cause session is not thread safe. The `session` property creates a thread-local session.   
Also a case for switching to httpx which afaik is thread safe and has the same interface as requests, that should let us do a direct subclass instead of a proxy

Reply:
OK I get it! @burnash so `Client` implements the same interface as `Session` but actually creates a Session per thread. so we should to drop `.session` here. please do a PR